### PR TITLE
client: Validate known deposit address after wallet connection

### LINKED ISSUE #879: BTC deposit address not in dex wallet
I sent a BTC deposit to the address shown in the GUI, but it didn't end up in the configured `dex` wallet.

```
# grep Deposit mainnet/logs/dexc.log
./mainnet/logs/dexc.log:2020-12-09 16:02:06.376 [INF] CORE: Loaded btc wallet configuration. Deposit address = bc1deadbeef....
```

Verifying whether the deposit address from `dexc` belongs to the `dex` wallet:
```
# bitcoin-cli -conf=/home/bitcoin/.bitcoin/bitcoin.conf getaddressinfo bc1deadbeef....
{
  "address": "bc1deadbeef....",
  "ismine": false,
  "solvable": false,
  "iswatchonly": false,
  "isscript": false,
  "iswitness": true,
  "ischange": false,
}
```

Confirming that it's the `dex` wallet that's open:

```
# bitcoin-cli -conf=/home/bitcoin/.bitcoin/bitcoin.conf getwalletinfo
{
  "walletname": "dex",
  "walletversion": 169900,
  "balance": 0.00000000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 0.00000000,
  "txcount": 12,
  "keypoololdest": 1606075542,
  "keypoolsize": 10000,
  "hdseedid": "xxxxxxx",
  "keypoolsize_hd_internal": 10000,
  "unlocked_until": 0,
  "paytxfee": 0.00000000,
  "private_keys_enabled": true,
  "avoid_reuse": false,
  "scanning": false
}
```

Any clues what might be the issue here?  The wallet was working correctly, a BTC withdrawal worked correctly yesterday. `bitcoind` is caught up to the tip as evidenced in `dexc.log`.

### PR DESCRIPTION
**Pull request by amassarwi**
This fixes #879 by validating deposit address for each wallet after successful connection.

I wiped a running wallet dir and created fresh one as #879 describes and the client
validated the deposit address and generated new one.

### CODE REVIEW

File: client/asset/btc/btc.go
Comment (chappjc) [inline client/asset/btc/btc.go:568]: wether -> whether.  But maybe rephrase like `ValidateAddress indicates if an address belongs to the wallet.`

File: client/asset/dcr/dcr.go
Comment (chappjc) [inline client/asset/dcr/dcr.go:397]: Maybe a quick doc string explaining why this is necessary (disambiguate the node and wallet methods).

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1071]: This should be uncommon, so I would not object to `Warnf`.  Also, since you're logging the old deposit address, you should log the new one too.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1072]: No need to shadow `err` with a new variable here.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1548]: Without having tested these changes, I'm confused about the `ReconfigureWallet` edit because it would seem to be requesting a new deposit address from the wallet backend *before* connecting to the new wallet.

But I'm thinking that with the changes to `connectWallet` that ensures the current deposit address are owned by the wallet, that `ReconfigureWallet` no longer even needs to unconditionally get a new deposit address.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1699]: I think this locking `Core.walletMtx` was a bug. `walletMtx` is meant to guard the `wallets` map.  The `xcWallet` has it's own mutex, which `setAddress` uses.

Review the use of `walletMtx` and the declaration of the mutex and the `wallets` field in the `Core` struct to see what I mean.  @buck54321 please confirm that there was not another consideration here, but it appears to be a misuse of `Core`'s `walletMtx` instead of the `xcWallet` mutex.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:2061]: The walletMtx was locked outside of the loop over `c.wallets` because it guards access to the `wallets` map.  It needs to go back that way for the loop.

But not the the loop starts goroutines, so the goroutines will fire off, the loop will complete, and `c.walletMtx.Unlock()` will allow each of the goroutines that need to lock it to lock it.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1072]: I'm also noting what appears to be an awkward use pattern that is making issues (or at least confusion with the walletMtx). `(*Core).NewDepositAddress(uint32)` was previously only consumed by the webserver api for generating new addresses, so it included a lookup of an `xcWallet` by id in the `wallets` map, thus necessitating locking `walletsMtx` for the map access.  However, here in `connectWallet` you already have the target `wcWallet`.  I would factor out the guts of `NewDepositAddress` into an unexported `newDepositAddress(w *xcWallet`)`, which you would use here.  The exported version would remain, doing a lookup by id and then calling the unexported one.

**Review by chappjc — CHANGES_REQUESTED** [review]

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:1072]: Sounds good, will do, thanks 👍

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:1699]: Oh, that's make sense, will update it, this might explain the deadlock I had when generating new address

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:1548]: Yep, that's true, but calling the `NewDepositAddress` inside the connectWallet needed the lock which caused 
a deadlock when generating new address, but as you described above, I would have already the xcWallet and can call the unexported method.

**Comment by amassarwi** [comment]
@chappjc Updated & tested, looks good & clean now, thanks for the patient review

**Comment by amassarwi** [comment]
@chappjc updated again 🔔

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1731]: Since we've defined `newDepositAddress` above to return an empty string for a non-nil error, it would be acceptable in this case to just do `return c.newDepositAddress(w)`, although generally it's wise to ensure return values work this way when dealing with exported methods from other packages.

Anyway, I've not actually tested any of this PR, so feel free to hold off for further comments.

Comment (amassarwi) [reply client/core/core.go:1731]: Right, can do

File: client/asset/interface.go
Comment (buck54321) [inline client/asset/interface.go:151]: An address can be valid without being "owned" by the wallet. Maybe `OwnsAddress` instead?

Comment (amassarwi) [reply client/asset/interface.go:151]: Agreed, in a sec.

File: client/asset/dcr/dcr.go
Comment (buck54321) [inline client/asset/dcr/dcr.go:558]: w**h**ether

File: client/asset/btc/walletclient.go
Comment (buck54321) [inline client/asset/btc/walletclient.go:229]: about **the** given address,

File: client/core/core.go
Comment (buck54321) [inline client/core/core.go:1056]: ~~after successful connection**. If the** deposit address...

**Review by buck54321 — CHANGES_REQUESTED** [review]
Running the ***simnet-setup.sh*** script, I'm getting an error

`error creating dcr wallet: decoded address is of unknown format: invalid format: version and/or checksum bytes missing`

**Comment by amassarwi** [comment]
@buck54321 updated mate, in CreateWallet the first address generation comes after the first connection, so I needed to skip the address validate in `connectWallet` if no address found for wallet.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1075]: The second `if` should be nested in the first.

```go
if addr != "" {
	mine, err := w.OwnsAddress(addr)
	if err != nil {
		return err
	}
	if !mine {
		nAddr, err := c.newDepositAddress(w)
```

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1067]: `w.address` being accessed here without RLock

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:1075]: opps

**Comment by amassarwi** [comment]
@chappjc Updated

**Review by JoeGruffins — APPROVED** [review]
Working for me.

**Comment by JoeGruffins** [comment]
I noticed while reviewing this that while the address is updated, if you change to another wallet with the same password, the balance is not updated until you restart the app. Also, changing a wallet mid trade could get you in trouble maybe? `[ERR] CORE: 127.0.0.1:17273 tick: {redeemMatches order e9f674a0e1419e53ffc958d6882985b263b64fcc33190087998e615d751ad152 - {error sending redeem transaction: -4: wallet.DumpWIFPrivateKey: item does not exist:: wallet.LoadPrivateKey: no address with id 121201201d26555503138db12cf615d2c26534ec (is wallet locked?)}}`

**Comment by amassarwi** [comment]
@JoeGruffins Thanks for the review mate, I noticed the balance issue as well and I think it's the same on master
and I was planning to create a separate issue/PR for that, does it make sense to fix this in a different PR ?

Regarding changing wallet in mid of trade no sure what would be the best approach here ?
We could maybe block block settings change while having a active matches ?
 
cc: @chappjc @buck54321

**Comment by chappjc** [comment]
@JoeGruffins What do you mean by "changing a wallet mid trade"?  Do you mean just a ReconfigureWallet or actually swapping out the actual wallet for one with entirely different keys and addresses?

We definitely should allow updating wallet settings at any time.  But if they actually change the underlying wallet, that's a no-no on the user's part.

**Comment by amassarwi** [comment]
@JoeGruffins @chappjc 
Created separated issue to update balance on wallet switch.
Will work on it after merging this one.

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:1503]: I'm not so sure this mutex change is desirable.  While there is the `(*Core).wallet` method that handles the locking, I think the reasoning for locking it up hear and keeping it locked until the new wallet is written into the map is to make the process atomic.  That is, we are guaranteed to be replacing oldWallet with the new one the old way, but now another goroutine could modify the map between this access and the write below.

**Review by chappjc — CHANGES_REQUESTED** [review]
Just the one remaining comment about the mutex in ReconfigureWallet, but also please review all the docs and ensure they read well and are properly punctuated.

changing BTC from gamma to default wallet:

[WRN] CORE: [btc]: Deposit address bcrt1q0ctryhtspw4zq3d6dgxsycqcq5ty66z405q9cm does not belong to connected wallet, generated new address: bcrt1q322tg0y2hzyp9zztr7d2twdclhqg88anvzxwwr

Changing DCR from beta to alpha account:

[WRN] CORE: [dcr]: Deposit address SsnbEmxCVXskgTHXvf3rEa17NA39qQuGHwQ does not belong to connected wallet, generated new address: SsoiAXujKA2aj3zsVVTNztfEPugwscyUrVz

reflected in web's deposit address:

image

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go]: Reverted

**Comment by amassarwi** [comment]
Updated 🔔

**Comment by JoeGruffins** [comment]
> What do you mean by "changing a wallet mid trade"? Do you mean just a ReconfigureWallet with the same wallet backend or actually swapping out the actual wallet for one with entirely different keys and addresses?

The second one. Isn't that what the user with the issue did, and the reason we need a different deposit address?

**Comment by chappjc** [comment]
@JoeGruffins yah, for sure the reason.  Just wanted to make sure that error wasn't possible with just a reconfigure during swaps.
This PR is only intended to facilitate wallet switches when there are no active swaps or orders. Perhaps we should put a warning to that effect on the wallet settings page.

**Comment by JoeGruffins** [comment]
At lease a warning would be prudent imo. You cannot change your backend wallet with active trades. Your funds locked in the order will not make it back to you.

**Review by chappjc — APPROVED** [review]
Looks good.  For a follow-up, I'll consider appropriate wording for a warning regarding changing the wallet to a different account/wallet with active swaps.

**Comment by amassarwi** [comment]
@chappjc 

I've pushed a placeholder:
image

Let me know when you a have final version in mind.

**Comment by JoeGruffins** [comment]
I think `might cause funds to be lost.` is better.

**Comment by JoeGruffins** [comment]
I notice that I see the warning: `[WRN] CORE: [dcr]: Deposit address SsdiLuQuNnjhpb8ezzgqfKmyRFPW5hQknsP does not belong to connected wallet, generated new address: SsfzvqzSByoRwb1htP7VDK15F5XSK2zb6np` once more upon the first restart and login with a different generated address. I don't see it on subsequent logins. So, I'm guessing it isn't being saved to the database when you first change it, but is upon login. Not a huge issue though...

**Comment by amassarwi** [comment]
@JoeGruffins Shouldn't regenerate on login, will double check

**Comment by amassarwi** [comment]
@JoeGruffins Updated in https://github.com/decred/dcrdex/pull/902/commits/67a53f7e3450bcf3ac9a41332f3f2f6c3bcdd915:

Needed to write `dbWallet` to db before calling `connectWallet`.   Before my last commit `connectWallet` would generate new address and write it to db but when it returns to `ReconfigureWallet` the `dbWallet` was saved to the db again overriding the new address with old address which caused another address generation on next login.

**Review by chappjc — CHANGES_REQUESTED** [review]

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:1567]: Valid concerns

**Comment by chappjc** [comment]
In fixing a wallet lock issue, I did some balance update work that adds the balance update in `ReconfigureWallet`. https://github.com/decred/dcrdex/pull/910/files#diff-13a4a6b12c1ae844ef90151638f194427cfdc545fceb58d52c0f6f33b81d859bR1556
However, I'll try to have this PR go in first to avoid making a conflict resolution job for you.

**Comment by amassarwi** [comment]
@chappjc great, will about better approach regarding https://github.com/decred/dcrdex/pull/902#discussion_r550345812

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go]: Updated to follow the solution I mentioned on matrix:

> chappjc: trying to figure out proper way to resolve https://github.com/decred/dcrdex/pull/902#pullrequestreview-560219940 - one way would be returning a bool from `connectWallet` something like: `generatedNewAddress` true/false.
then i could check that flag in `ReconfigureWallet` after calling connectWallet - if true I can store xcWallet's (new) address as dbWallet.Address before updating the dbwallet in `ReconfigureWallet`

**Comment by amassarwi** [comment]
@chappjc updated according to https://github.com/decred/dcrdex/pull/902#discussion_r550827492

**Pull request merged**
